plugins/catalog-backend/src/ingestion/processors/CodeOwnersProcessor.ts

```diff
@@ -0,0 +1,172 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity, LocationSpec } from '@backstage/catalog-model';
+import {
+  LocationProcessor,
+  LocationProcessorEmit,
+  LocationProcessorRead,
+} from './types';
+import * as codeowners from 'codeowners-utils';
+import { CodeOwnersEntry } from 'codeowners-utils';
+import parseGitUri from 'git-url-parse';
+import { filter, head, get, pipe, reverse } from 'lodash/fp';
+
+// NOTE: This can be removed when ES2021 is implemented
+import 'core-js/features/promise';
+
+const ALLOWED_LOCATION_TYPES = [
+  'azure/api',
+  'bitbucket/api',
+  'github',
+  'github/api',
+  'gitlab',
+  'gitlab/api',
+];
+
+export class CodeOwnersProcessor implements LocationProcessor {
+  async processEntity(
+    entity: Entity,
+    location: LocationSpec,
+    _emit: LocationProcessorEmit,
+    read: LocationProcessorRead,
+  ): Promise<Entity> {
+    // Only continue if the owner is not set
+    if (
+      !entity ||
+      !['Component', 'API'].includes(entity.kind) ||
+      !ALLOWED_LOCATION_TYPES.includes(location.type) ||
+      (entity.spec && entity.spec.owner)
+    ) {
+      return entity;
+    }
+
+    const owner = await resolveCodeOwner(location, read);
+
+    return {
+      ...entity,
+      spec: { ...entity.spec, owner },
+    };
+  }
+}
+
+export async function resolveCodeOwner(
+  location: LocationSpec,
+  read: LocationProcessorRead,
+): Promise<string | undefined> {
+  const ownersText = await findRawCodeOwners(location, read);
+
+  if (!ownersText) {
+    throw Error(`Unable to find codeowners file for: ${location.target}`);
+  }
+
+  const owners = parseCodeOwners(ownersText);
+
+  return findPrimaryCodeOwner(owners);
+}
+
+export async function findRawCodeOwners(
+  location: LocationSpec,
+  read: LocationProcessorRead,
+): Promise<string | undefined> {
+  const readOwnerLocation = async (basePath: string): Promise<string> => {
+    const ownerLocation = buildCodeOwnerLocation(
+      location,
+      `${basePath}/CODEOWNERS`,
+    );
+
+    return readLocation(ownerLocation, read);
+  };
+
+  const gitProvider = location.type.split('/')[0];
+
+  return Promise.any([
+    readOwnerLocation(`/.${gitProvider}`),
+    readOwnerLocation(''),
+    readOwnerLocation('/docs'),
+  ]);
+}
+
+export function parseCodeOwners(ownersText: string) {
+  return codeowners.parse(ownersText);
+}
+
+export function findPrimaryCodeOwner(
+  owners: CodeOwnersEntry[],
+  pattern = '*',
+): string | undefined {
+  return pipe(
+    filter((e: CodeOwnersEntry) => e.pattern === pattern),
+    reverse,
+    head,
+    get('owners'),
+    head,
```

Comment:
I'm also wondering if we should remove the `@` suffix? My thinking we will want the owner to ideally map to a `Group` entity. I can't recall the restrictions on `metadata.name` of hand.

Reply:
Yes, we will have to add a matching step at the end somehow, that takes whatever came out of the codeowners file's right hand column and finds a matching Group or User entity for it. The matching will most likely be via looking for certain annotations, such as `github.com/email` etc.

Since all of this isn't 100% in place yet, maybe we can skip the catalog matching for now. I'd be happy to just for example do

- if it's on the form of an email, take the part before the @
- if it's on the form of a team, remove the @org/ part and keep only the team name

and call it a day for now

Reply:
Yep, see [this](https://github.com/spotify/backstage/pull/2613/files#diff-6f77deb3b70e47a6c457a0573782eb3bR152) for the lookup bits. It's more tricky to do during processing though, as we may not actually have those user entities in the catalog yet. Tbh we should likely make sure that org data is always processed first and is available for lookups in any other steps.

+1 for not solving that now :grin:

Reply:
I've added the function `normalizeCodeOwner` to handle this case for now